### ISSUE TITLE
Mypy api signature overly specific

### ISSUE THREAD

**Issue by atagar** [python/mypy#8693]
Hi mypy. Very low priority inquiry from a newcomer.

Python [advises that arguments types should be abstract](https://docs.python.org/3/library/typing.html#typing.List) (**Sequence** rather than **List**). MyPy's documentation [advises the same](https://mypy.readthedocs.io/en/stable/common_issues.html#incompatible-overrides).

However, mypy itself [uses specific types for its own API](https://github.com/python/mypy/blob/master/mypy/api.py#L66). As a result my code that uses it...

```
def type_issues(args: Sequence[str]) -> Dict[str, List['stem.util.test_tools.Issue']]:
  lines = mypy.api.run(args)[0].splitlines()
  ... etc...
```

Fails its mypy checks with...

```
Argument 1 to "run" has incompatible type "Sequence[str]"; expected "List[str]"  [arg-type]
```

The aforementioned mypy documentation says that mypy should detect this issue within itself...

> mypy won’t report an error for OverwriteReturnSpecific but it does for OverwriteReturnGeneric and OverwriteArgumentSpecific

Does mypy pass its own checks?

Thanks! And apologies if I'm missing something. Again, I'm green to the wonderful world of python type checking so delighted to be corrected.

**Comment by Michael0x2a** [comment]
> Python advises that arguments types should be abstract (Sequence rather than List). MyPy's documentation advises the same.

The two pages you're linking to are actually discussing two different principles.

The advice given in the Python docs is basically a variation of something known as the "[Robustness principle](https://en.wikipedia.org/wiki/Robustness_principle)" which states that we should write code that's liberal in what it accepts and conservative in what it receives.

For example, `List[T]` is a pretty specific type: it can only accept builtin lists (or subclasses of list). This isn't necessarily wrong: if your function relies on the input being a list, this type signature would be correct.

But if your function actually isn't this restrictive and if you want to make this function available for widespread use, it'd be better to type it to accept a more broader type such as `Sequence[T]`. This way, the caller could also pass in other sequence-like types such as tuples without needing to do extra work.

The degree to which people follow this principle does vary. Trying to type everything as generally as possible isn't always the best use of time, and it can sometimes be better to just use the concrete type and move on. It's also easier to widen a type signature after-the-fact then it is to narrow it -- it's easier to give then to take away. Basically, starting with a too-general API runs the risk of tying your hands down the line.

The section of the mypy docs you linked to is discussing a completely different principle known as the "[Liskov substitution principle](https://en.wikipedia.org/wiki/Liskov_substitution_principle)". In short, if you have a type A that is a subtype of another class B, it must always be possible to *substitute* and instances of B with instances of A without breaking anything.

For example, suppose you create a custom sequence object that subclasses `Sequence[T]`. You need to make sure your custom object still behaves like a Sequence in all regards to avoid introducing subtle (or not-so-subtle) bugs. For example, if we break the principle and do something like this:

    class MySequence(Sequence[T]):
        def __len__(self, boom: str) -> str:
            time.sleep(1000)
            return "Nobody expects the spanish inquisition"

...this would clearly break any downstream code that expects `__len__()` actually returning a length.

Mypy partially enforces this principle by mandating that any methods you override in a subclass must be compatible with the original parent signature. The section of the doc you're looking at is explaining what rules a overridden method must follow to remain compatible with the parent method signature.

However, this principle isn't relevant to the mypy.api.run function: there's no subtyping going on anywhere, so the Liskov substitution principle isn't relevant.

> However, mypy itself uses specific types for its own API.

In this case, it's because the `args` parameter is supposed override the [`sys.argv`](https://docs.python.org/3/library/sys.html#sys.argv), which is a list of arguments passed in from the command line. And since `sys.argv` has type list, the easiest thing to do was to just give `args` the same type.

Tweaking mypy.api.run to accept a Sequence instead of a List is probably a reasonable idea, albeit a low-priority one.

> Does mypy pass its own checks?

It does, yes.

**Comment by atagar** [comment]
Ahhh! Thank you Michael for that phenomenal explanation, I appreciate it.

Changing the run method signature would be such a trivially low priority I doubt it merits a ticket so closing this.
